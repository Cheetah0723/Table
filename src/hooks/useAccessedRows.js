import { useMemo } from 'react'

export default function useAccessedRows ({
  debug, data, columns, subRowsKey,
}) {
  return useMemo(
    () => {
      if (debug) console.info('getAccessedRows')

      // Access the row's data
      const accessRow = (data, i, depth = 0) => {
        // Keep the original reference around
        const original = data

        // Process any subRows
        const subRows = data[subRowsKey]
          ? data[subRowsKey].map((d, i) => accessRow(d, i, depth + 1))
          : undefined

        const row = {
          original,
          index: i,
          subRows,
          depth,
        }

        // Create the cells and values
        row.values = {}
        columns.forEach(column => {
          row.values[column.id] = column.accessor ? column.accessor(data) : undefined
        })

        return row
      }

      // Use the resolved data
      return data.map((d, i) => accessRow(d, i))
    },
    [data, columns]
  )
}
