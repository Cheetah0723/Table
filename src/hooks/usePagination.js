import { useMemo, useLayoutEffect } from 'react'
import PropTypes from 'prop-types'

//
import { addActions, actions } from '../actions'
import { defaultState } from './useTableState'

defaultState.pageSize = 10
defaultState.pageIndex = 0

addActions({
  pageChange: '__pageChange__',
  pageSizeChange: '__pageSizeChange__',
})

const propTypes = {
  // General
  manualPagination: PropTypes.bool,
}

export const usePagination = props => {
  PropTypes.checkPropTypes(propTypes, props, 'property', 'usePagination')

  const {
    rows,
    manualPagination,
    disablePageResetOnDataChange,
    debug,
    state: [
      {
        pageSize,
        pageIndex,
        pageCount: userPageCount,
        filters,
        groupBy,
        sortBy,
      },
      setState,
    ],
  } = props

  const pageOptions = useMemo(
    () => [...new Array(userPageCount)].map((d, i) => i),
    [userPageCount]
  )

  const rowDep = disablePageResetOnDataChange ? null : rows

  useLayoutEffect(() => {
    setState(
      old => ({
        ...old,
        pageIndex: 0,
      }),
      actions.pageChange
    )
  }, [setState, rowDep, filters, groupBy, sortBy])

  const { pages, pageCount } = useMemo(() => {
    if (manualPagination) {
      return {
        pages: [rows],
        pageCount: userPageCount,
      }
    }
    if (debug) console.info('getPages')

    // Create a new pages with the first page ready to go.
    const pages = rows.length ? [] : [[]]

    // Start the pageIndex and currentPage cursors
    let cursor = 0
    while (cursor < rows.length) {
      const end = cursor + pageSize
      pages.push(rows.slice(cursor, end))
      cursor = end
    }

    const pageCount = pages.length

    return {
      pages,
      pageCount,
      pageOptions,
    }
  }, [manualPagination, debug, rows, pageOptions, userPageCount, pageSize])

  const page = manualPagination ? rows : pages[pageIndex] || []
  const canPreviousPage = pageIndex > 0
  const canNextPage = pageIndex < pageCount - 1

  const gotoPage = pageIndex => {
    if (debug) console.info('gotoPage')
    return setState(old => {
      if (pageIndex < 0 || pageIndex > pageCount - 1) {
        return old
      }
      return {
        ...old,
        pageIndex,
      }
    }, actions.pageChange)
  }

  const previousPage = () => {
    return gotoPage(pageIndex - 1)
  }

  const nextPage = () => {
    return gotoPage(pageIndex + 1)
  }

  const setPageSize = pageSize => {
    setState(old => {
      const topRowIndex = old.pageSize * old.pageIndex
      const pageIndex = Math.floor(topRowIndex / pageSize)
      return {
        ...old,
        pageIndex,
        pageSize,
      }
    }, actions.pageSizeChange)
  }

  return {
    ...props,
    pages,
    pageOptions,
    pageCount,
    page,
    canPreviousPage,
    canNextPage,
    gotoPage,
    previousPage,
    nextPage,
    setPageSize,
    pageIndex,
    pageSize,
  }
}
