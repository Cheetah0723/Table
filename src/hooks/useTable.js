import React from 'react'
import PropTypes from 'prop-types'
//
import {
  applyHooks,
  applyPropHooks,
  mergeProps,
  flexRender,
  decorateColumnTree,
  makeHeaderGroups,
  flattenBy,
  determineHeaderVisibility,
} from '../utils'

const propTypes = {
  // General
  data: PropTypes.array.isRequired,
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  defaultColumn: PropTypes.object,
  getSubRows: PropTypes.func,
  getRowID: PropTypes.func,
  debug: PropTypes.bool,
}

const renderErr =
  'You must specify a valid render component. This could be "column.Cell", "column.Header", "column.Filter", "column.Aggregated" or any other custom renderer component.'

export const defaultState = {}

const defaultInitialState = {}
const defaultColumnInstance = {}
const defaultReducer = (old, newState) => newState
const defaultGetSubRows = (row, index) => row.subRows || []
const defaultGetRowID = (row, index) => index

export const useTable = (props, ...plugins) => {
  // Validate props
  PropTypes.checkPropTypes(propTypes, props, 'property', 'useTable')

  // Destructure props
  let {
    data,
    columns: userColumns,
    initialState = defaultInitialState,
    state: userState,
    defaultColumn = defaultColumnInstance,
    getSubRows = defaultGetSubRows,
    getRowID = defaultGetRowID,
    reducer = defaultReducer,
    debug,
  } = props

  debug = process.env.NODE_ENV === 'production' ? false : debug

  // But use the users table state if provided
  let [originalState, originalSetState] = React.useState({
    ...defaultState,
    ...initialState,
  })

  const state = React.useMemo(() => {
    if (userState) {
      const newState = {
        ...originalState,
      }
      Object.keys(userState).forEach(key => {
        newState[key] = userState[key]
      })
      return newState
    }
    return originalState
  }, [originalState, userState])

  const setState = React.useCallback(
    (updater, type) => {
      return originalSetState(old => {
        const newState = typeof updater === 'function' ? updater(old) : updater
        return reducer(old, newState, type)
      })
    },
    [reducer]
  )

  // The table instance ref
  let instanceRef = React.useRef({})

  Object.assign(instanceRef.current, {
    ...props,
    data, // The raw data
    state,
    setState, // The resolved table state
    plugins, // All resolved plugins
    hooks: {
      columnsBeforeHeaderGroups: [],
      columnsBeforeHeaderGroupsDeps: [],
      useBeforeDimensions: [],
      useMain: [],
      useRows: [],
      prepareRow: [],
      getTableProps: [],
      getTableBodyProps: [],
      getRowProps: [],
      getHeaderGroupProps: [],
      getHeaderProps: [],
      getCellProps: [],
    },
  })

  // Allow plugins to register hooks
  if (process.env.NODE_ENV === 'development' && debug) console.time('plugins')

  plugins.filter(Boolean).forEach(plugin => {
    plugin(instanceRef.current.hooks)
  })

  if (process.env.NODE_ENV === 'development' && debug)
    console.timeEnd('plugins')

  // Decorate All the columns
  let columns = React.useMemo(
    () => decorateColumnTree(userColumns, defaultColumn),
    [defaultColumn, userColumns]
  )

  // Get the flat list of all columns andllow hooks to decorate
  // those columns (and trigger this memoization via deps)
  let flatColumns = React.useMemo(() => {
    if (process.env.NODE_ENV === 'development' && debug)
      console.time('hooks.columnsBeforeHeaderGroups')

    let newColumns = applyHooks(
      instanceRef.current.hooks.columnsBeforeHeaderGroups,
      flattenBy(columns, 'columns'),
      instanceRef.current
    )

    if (process.env.NODE_ENV === 'development' && debug)
      console.timeEnd('hooks.columnsBeforeHeaderGroups')
    return newColumns
  }, [
    columns,
    debug,
    // eslint-disable-next-line react-hooks/exhaustive-deps
    ...applyHooks(
      instanceRef.current.hooks.columnsBeforeHeaderGroupsDeps,
      [],
      instanceRef.current
    ),
  ])

  // Make the headerGroups
  const headerGroups = React.useMemo(
    () => makeHeaderGroups(flatColumns, defaultColumn),
    [defaultColumn, flatColumns]
  )

  const headers = React.useMemo(() => headerGroups[0].headers, [headerGroups])

  Object.assign(instanceRef.current, {
    columns,
    flatColumns,
    headerGroups,
    headers,
  })

  // Access the row model
  const [rows, flatRows] = React.useMemo(() => {
    if (process.env.NODE_ENV === 'development' && debug)
      console.time('getAccessedRows')

    let flatRows = []

    // Access the row's data
    const accessRow = (originalRow, i, depth = 0, parentPath = []) => {
      // Keep the original reference around
      const original = originalRow

      const rowID = getRowID(originalRow, i)

      // Make the new path for the row
      const path = [...parentPath, rowID]

      const row = {
        original,
        index: i,
        path, // used to create a key for each row even if not nested
        depth,
        cells: [{}], // This is a dummy cell
      }

      flatRows.push(row)

      // Process any subRows
      let subRows = getSubRows(originalRow, i)

      if (subRows) {
        row.subRows = subRows.map((d, i) => accessRow(d, i, depth + 1, path))
      }

      // Override common array functions (and the dummy cell's getCellProps function)
      // to show an error if it is accessed without calling prepareRow
      const unpreparedAccessWarning = () => {
        throw new Error(
          'React-Table: You have not called prepareRow(row) one or more rows you are attempting to render.'
        )
      }
      row.cells.map = unpreparedAccessWarning
      row.cells.filter = unpreparedAccessWarning
      row.cells.forEach = unpreparedAccessWarning
      row.cells[0].getCellProps = unpreparedAccessWarning

      // Create the cells and values
      row.values = {}
      flatColumns.forEach(column => {
        row.values[column.id] = column.accessor
          ? column.accessor(originalRow, i, { subRows, depth, data })
          : undefined
      })

      return row
    }

    // Use the resolved data
    const accessedData = data.map((d, i) => accessRow(d, i))
    if (process.env.NODE_ENV === 'development' && debug)
      console.timeEnd('getAccessedRows')
    return [accessedData, flatRows]
  }, [debug, data, getRowID, getSubRows, flatColumns])

  instanceRef.current.rows = rows
  instanceRef.current.flatRows = flatRows

  // Determine column visibility
  determineHeaderVisibility(instanceRef.current)

  // Provide a flat header list for utilities
  instanceRef.current.flatHeaders = headerGroups.reduce(
    (all, headerGroup) => [...all, ...headerGroup.headers],
    []
  )

  if (process.env.NODE_ENV === 'development' && debug)
    console.time('hooks.useBeforeDimensions')
  instanceRef.current = applyHooks(
    instanceRef.current.hooks.useBeforeDimensions,
    instanceRef.current
  )
  if (process.env.NODE_ENV === 'development' && debug)
    console.timeEnd('hooks.useBeforeDimensions')

  calculateDimensions(instanceRef.current)

  if (process.env.NODE_ENV === 'development' && debug)
    console.time('hooks.useMain')
  instanceRef.current = applyHooks(
    instanceRef.current.hooks.useMain,
    instanceRef.current
  )
  if (process.env.NODE_ENV === 'development' && debug)
    console.timeEnd('hooks.useMain')

  // Each materialized header needs to be assigned a render function and other
  // prop getter properties here.
  instanceRef.current.flatHeaders.forEach(column => {
    // Give columns/headers rendering power
    column.render = (type, userProps = {}) => {
      const Comp = typeof type === 'string' ? column[type] : type

      if (typeof Comp === 'undefined') {
        throw new Error(renderErr)
      }

      return flexRender(Comp, {
        ...instanceRef.current,
        column,
        ...userProps,
      })
    }

    // Give columns/headers a default getHeaderProps
    column.getHeaderProps = props =>
      mergeProps(
        {
          key: ['header', column.id].join('_'),
          colSpan: column.totalVisibleHeaderCount,
        },
        applyPropHooks(
          instanceRef.current.hooks.getHeaderProps,
          column,
          instanceRef.current
        ),
        props
      )
  })

  instanceRef.current.headerGroups.forEach((headerGroup, i) => {
    // Filter out any headers and headerGroups that don't have visible columns
    headerGroup.headers = headerGroup.headers.filter(header => {
      const recurse = headers =>
        headers.filter(header => {
          if (header.headers) {
            return recurse(header.headers)
          }
          return header.isVisible
        }).length
      if (header.headers) {
        return recurse(header.headers)
      }
      return header.isVisible
    })

    // Give headerGroups getRowProps
    if (headerGroup.headers.length) {
      headerGroup.getHeaderGroupProps = (props = {}) =>
        mergeProps(
          {
            key: [`header${i}`].join('_'),
          },
          applyPropHooks(
            instanceRef.current.hooks.getHeaderGroupProps,
            headerGroup,
            instanceRef.current
          ),
          props
        )

      return true
    }
  })

  // Run the rows (this could be a dangerous hook with a ton of data)
  if (process.env.NODE_ENV === 'development' && debug)
    console.time('hooks.useRows')
  instanceRef.current.rows = applyHooks(
    instanceRef.current.hooks.useRows,
    instanceRef.current.rows,
    instanceRef.current
  )
  if (process.env.NODE_ENV === 'development' && debug)
    console.timeEnd('hooks.useRows')

  // The prepareRow function is absolutely necessary and MUST be called on
  // any rows the user wishes to be displayed.

  instanceRef.current.prepareRow = React.useCallback(row => {
    row.getRowProps = props =>
      mergeProps(
        { key: ['row', ...row.path].join('_') },
        applyPropHooks(
          instanceRef.current.hooks.getRowProps,
          row,
          instanceRef.current
        ),
        props
      )

    // Build the visible cells for each row
    row.cells = instanceRef.current.flatColumns
      .filter(d => d.isVisible)
      .map(column => {
        const cell = {
          column,
          row,
          value: row.values[column.id],
        }

        // Give each cell a getCellProps base
        cell.getCellProps = props => {
          const columnPathStr = [...row.path, column.id].join('_')
          return mergeProps(
            {
              key: ['cell', columnPathStr].join('_'),
            },
            applyPropHooks(
              instanceRef.current.hooks.getCellProps,
              cell,
              instanceRef.current
            ),
            props
          )
        }

        // Give each cell a renderer function (supports multiple renderers)
        cell.render = (type, userProps = {}) => {
          const Comp = typeof type === 'string' ? column[type] : type

          if (typeof Comp === 'undefined') {
            throw new Error(renderErr)
          }

          return flexRender(Comp, {
            ...instanceRef.current,
            column,
            row,
            cell,
            ...userProps,
          })
        }

        return cell
      })

    // need to apply any row specific hooks (useExpanded requires this)
    applyHooks(instanceRef.current.hooks.prepareRow, row, instanceRef.current)
  }, [])

  instanceRef.current.getTableProps = userProps =>
    mergeProps(
      applyPropHooks(
        instanceRef.current.hooks.getTableProps,
        instanceRef.current
      ),
      userProps
    )

  instanceRef.current.getTableBodyProps = userProps =>
    mergeProps(
      applyPropHooks(
        instanceRef.current.hooks.getTableBodyProps,
        instanceRef.current
      ),
      userProps
    )

  return instanceRef.current
}

function calculateDimensions(instance) {
  const { headers } = instance

  instance.totalColumnsWidth = calculateHeaderWidths(headers)
}

function calculateHeaderWidths(headers, left = 0) {
  let sumTotalWidth = 0

  headers.forEach(header => {
    let { headers: subHeaders } = header

    header.totalLeft = left

    if (subHeaders && subHeaders.length) {
      header.totalWidth = calculateHeaderWidths(subHeaders, left)
    } else {
      header.totalWidth = Math.min(
        Math.max(header.minWidth, header.width),
        header.maxWidth
      )
    }
    if (header.isVisible) {
      left += header.totalWidth
      sumTotalWidth += header.totalWidth
    }
  })

  return sumTotalWidth
}
