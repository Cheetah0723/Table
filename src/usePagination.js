import { useMemo, useState, useLayoutEffect } from 'react'
import PropTypes from 'prop-types'

import { warnUnknownProps } from './utils'
import useControlledState from './hooks/useControlledState'

const propTypes = {
  defaultPageSize: PropTypes.number,
  defaultPageIndex: PropTypes.number,
  pageSize: PropTypes.number,
  pages: PropTypes.number,
  pageIndex: PropTypes.number,
  onStateChange: PropTypes.func,
  stateReducer: PropTypes.func,
  subRowsKey: PropTypes.string,
  debug: PropTypes.bool,
}

export const actions = {
  pageChange: '__pageChange__',
}

export default function usePagination (
  {
    debug: parentDebug, rows, prepareRows, groupedRows, sortedRows, filteredRows,
  },
  props = {}
) {
  // Validate props
  PropTypes.checkPropTypes(propTypes, props, 'property', 'usePagination')

  // Destructure props
  const {
    defaultPageSize = 10,
    defaultPageIndex = 0,
    pageSize: userPageSize,
    pageIndex: userPageIndex,
    state: userState,
    manualPagination,
    pageCount: userPageCount,
    pageOptions: userPageOptions,
    debug = parentDebug,
    ...rest
  } = props

  warnUnknownProps(rest)

  const defaultState = useState({})
  const localState = userState || defaultState

  // Build the controllable state
  const [{ pageSize, pageIndex }, setState] = useControlledState(
    localState,
    {
      pageSize: defaultPageSize,
      pageIndex: defaultPageIndex,
    },
    {
      pageSize: userPageSize,
      pageIndex: userPageIndex,
    }
  )

  useLayoutEffect(
    () => {
      if (manualPagination) {
        return
      }
      setState(
        old => ({
          ...old,
          pageIndex: 0,
        }),
        actions.pageChange
      )
    },
    [groupedRows, sortedRows, filteredRows]
  )

  const { pages, pageCount } = useMemo(
    () => {
      if (manualPagination) {
        return {
          pages: [rows],
          pageCount: userPageCount,
        }
      }
      if (debug) console.info('getPages')

      // Create a new pages with the first page ready to go.
      const pages = rows.length ? [] : [[]]

      // Start the pageIndex and currentPage cursors
      let cursor = 0
      while (cursor < rows.length) {
        const end = cursor + pageSize
        pages.push(rows.slice(cursor, end))
        cursor = end
      }

      const pageCount = pages.length

      return {
        pages,
        pageCount,
        pageOptions,
      }
    },
    [rows, pageSize, userPageCount]
  )

  const pageOptions = [...new Array(pageCount)].map((d, i) => i)
  const page = manualPagination ? rows : pages[pageIndex] || []
  const canPreviousPage = pageIndex > 0
  const canNextPage = pageIndex < pageCount - 1

  const gotoPage = pageIndex => {
    if (debug) console.info('gotoPage')
    return setState(old => {
      if (pageIndex < 0 || pageIndex > pageCount - 1) {
        return old
      }
      return {
        ...old,
        pageIndex,
      }
    }, actions.pageChange)
  }

  const previousPage = () => gotoPage(pageIndex - 1)

  const nextPage = () => gotoPage(pageIndex + 1)

  const setPageSize = pageSize => {
    setState(old => ({
      ...old,
      pageSize,
    }))
  }

  prepareRows(page)

  return {
    pages,
    pageIndex,
    pageOptions,
    page,
    canPreviousPage,
    canNextPage,
    gotoPage,
    previousPage,
    nextPage,
    setPageSize,
  }
}
