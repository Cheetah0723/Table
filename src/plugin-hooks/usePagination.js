import React from 'react'
import PropTypes from 'prop-types'

//
import { addActions, actions } from '../actions'
import { defaultState } from '../hooks/useTable'
import { ensurePluginOrder, safeUseLayoutEffect, expandRows } from '../utils'

defaultState.pageSize = 10
defaultState.pageIndex = 0

addActions('pageChange', 'pageSizeChange')

const propTypes = {
  // General
  manualPagination: PropTypes.bool,
  paginateExpandedRows: PropTypes.bool,
}

export const usePagination = hooks => {
  hooks.useMain.push(useMain)
}

usePagination.pluginName = 'usePagination'

function useMain(instance) {
  PropTypes.checkPropTypes(propTypes, instance, 'property', 'usePagination')

  const {
    data,
    rows,
    manualPagination,
    disablePageResetOnDataChange,
    manualExpandedKey = 'expanded',
    debug,
    plugins,
    pageCount: userPageCount,
    paginateExpandedRows = true,
    expandSubRows = true,
    state: { pageSize, pageIndex, filters, groupBy, sortBy, expanded },
    setState,
  } = instance

  ensurePluginOrder(
    plugins,
    ['useFilters', 'useGroupBy', 'useSortBy', 'useExpanded'],
    'usePagination',
    []
  )

  const rowDep = manualPagination ? null : data

  const isPageIndexMountedRef = React.useRef()

  // Bypass any effects from firing when this changes
  const disablePageResetOnDataChangeRef = React.useRef()
  disablePageResetOnDataChangeRef.current = disablePageResetOnDataChange

  safeUseLayoutEffect(() => {
    if (
      isPageIndexMountedRef.current &&
      !disablePageResetOnDataChangeRef.current
    ) {
      setState(
        old => ({
          ...old,
          pageIndex: 0,
        }),
        actions.pageChange
      )
    }
    isPageIndexMountedRef.current = true
  }, [setState, rowDep, filters, groupBy, sortBy])

  const pageCount = manualPagination
    ? userPageCount
    : Math.ceil(rows.length / pageSize)

  const pageOptions = React.useMemo(
    () => (pageCount > 0 ? [...new Array(pageCount)].map((d, i) => i) : []),
    [pageCount]
  )

  const page = React.useMemo(() => {
    let page

    if (manualPagination) {
      page = rows
    } else {
      if (process.env.NODE_ENV === 'development' && debug)
        console.info('getPage')

      const pageStart = pageSize * pageIndex
      const pageEnd = pageStart + pageSize

      page = rows.slice(pageStart, pageEnd)
    }

    if (paginateExpandedRows) {
      return page
    }

    return expandRows(page, { manualExpandedKey, expanded, expandSubRows })
  }, [
    debug,
    expandSubRows,
    expanded,
    manualExpandedKey,
    manualPagination,
    pageIndex,
    pageSize,
    paginateExpandedRows,
    rows,
  ])

  const canPreviousPage = pageIndex > 0
  const canNextPage = pageCount === -1 || pageIndex < pageCount - 1

  const gotoPage = React.useCallback(
    updater => {
      if (process.env.NODE_ENV === 'development' && debug)
        console.info('gotoPage')
      return setState(old => {
        const newPageIndex =
          typeof updater === 'function' ? updater(old.pageIndex) : updater

        if (newPageIndex < 0 || newPageIndex > pageCount - 1) {
          return old
        }
        return {
          ...old,
          pageIndex: newPageIndex,
        }
      }, actions.pageChange)
    },
    [debug, pageCount, setState]
  )

  const previousPage = React.useCallback(() => {
    return gotoPage(old => old - 1)
  }, [gotoPage])

  const nextPage = React.useCallback(() => {
    return gotoPage(old => old + 1)
  }, [gotoPage])

  const setPageSize = React.useCallback(
    pageSize => {
      setState(old => {
        const topRowIndex = old.pageSize * old.pageIndex
        const pageIndex = Math.floor(topRowIndex / pageSize)
        return {
          ...old,
          pageIndex,
          pageSize,
        }
      }, actions.pageSizeChange)
    },
    [setState]
  )

  return {
    ...instance,
    pageOptions,
    pageCount,
    page,
    canPreviousPage,
    canNextPage,
    gotoPage,
    previousPage,
    nextPage,
    setPageSize,
    pageIndex,
    pageSize,
  }
}
