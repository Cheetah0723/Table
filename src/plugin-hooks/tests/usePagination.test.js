import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { useTable } from '../../hooks/useTable'
import { usePagination } from '../usePagination'

const data = [...new Array(100)].map((d, i) => ({
  firstName: 'tanner ' + (i + 1),
  lastName: 'linsley',
  age: 29,
  visits: 100,
  status: 'In Relationship',
  progress: 50,
}))

function Table({ columns, data }) {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    prepareRow,
    page,
    canPreviousPage,
    canNextPage,
    pageOptions,
    pageCount,
    gotoPage,
    nextPage,
    previousPage,
    setPageSize,
    state: { pageIndex, pageSize },
  } = useTable(
    {
      columns,
      data,
      initialState: { pageIndex: 2 },
    },
    usePagination
  )

  return (
    <>
      <table {...getTableProps()}>
        <thead>
          {headerGroups.map(headerGroup => (
            <tr {...headerGroup.getHeaderGroupProps()}>
              {headerGroup.headers.map(column => (
                <th {...column.getHeaderProps()}>{column.render('Header')}</th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {page.map(
            (row, i) =>
              prepareRow(row) || (
                <tr {...row.getRowProps()}>
                  {row.cells.map(cell => {
                    return (
                      <td {...cell.getCellProps()}>{cell.render('Cell')}</td>
                    )
                  })}
                </tr>
              )
          )}
        </tbody>
      </table>
      <div className="pagination">
        <button onClick={() => gotoPage(0)} disabled={!canPreviousPage}>
          {'<<'}
        </button>{' '}
        <button onClick={() => previousPage()} disabled={!canPreviousPage}>
          {'<'}
        </button>{' '}
        <button onClick={() => nextPage()} disabled={!canNextPage}>
          {'>'}
        </button>{' '}
        <button onClick={() => gotoPage(pageCount - 1)} disabled={!canNextPage}>
          {'>>'}
        </button>{' '}
        <span>
          Page{' '}
          <strong>
            {pageIndex + 1} of {pageOptions.length}
          </strong>{' '}
        </span>
        <span>
          | Go to page:{' '}
          <input
            type="number"
            defaultValue={pageIndex + 1}
            onChange={e => {
              const page = e.target.value ? Number(e.target.value) - 1 : 0
              gotoPage(page)
            }}
            style={{ width: '100px' }}
          />
        </span>{' '}
        <select
          value={pageSize}
          onChange={e => {
            setPageSize(Number(e.target.value))
          }}
        >
          {[10, 20, 30, 40, 50].map(pageSize => (
            <option key={pageSize} value={pageSize}>
              Show {pageSize}
            </option>
          ))}
        </select>
      </div>
    </>
  )
}

function App() {
  const columns = React.useMemo(
    () => [
      {
        Header: 'Name',
        columns: [
          {
            Header: 'First Name',
            accessor: 'firstName',
          },
          {
            Header: 'Last Name',
            accessor: 'lastName',
          },
        ],
      },
      {
        Header: 'Info',
        columns: [
          {
            Header: 'Age',
            accessor: 'age',
          },
          {
            Header: 'Visits',
            accessor: 'visits',
          },
          {
            Header: 'Status',
            accessor: 'status',
          },
          {
            Header: 'Profile Progress',
            accessor: 'progress',
          },
        ],
      },
    ],
    []
  )

  return <Table columns={columns} data={data} />
}

test('renders a paginated table', () => {
  const { getByText, asFragment } = render(<App />)

  const fragment1 = asFragment()

  fireEvent.click(getByText('>'))

  const fragment2 = asFragment()

  fireEvent.click(getByText('>'))

  const fragment3 = asFragment()

  fireEvent.click(getByText('>>'))

  const fragment4 = asFragment()

  expect(fragment1).toMatchDiffSnapshot(fragment2)
  expect(fragment2).toMatchDiffSnapshot(fragment3)
  expect(fragment3).toMatchDiffSnapshot(fragment4)
})
