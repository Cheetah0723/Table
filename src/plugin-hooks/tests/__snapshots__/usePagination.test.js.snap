// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a paginated table 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -67,11 +67,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 21
+           tanner 31
          </td>
          <td
            class=""
          >
            linsley
@@ -101,11 +101,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 22
+           tanner 32
          </td>
          <td
            class=""
          >
            linsley
@@ -135,11 +135,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 23
+           tanner 33
          </td>
          <td
            class=""
          >
            linsley
@@ -169,11 +169,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 24
+           tanner 34
          </td>
          <td
            class=""
          >
            linsley
@@ -203,11 +203,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 25
+           tanner 35
          </td>
          <td
            class=""
          >
            linsley
@@ -237,11 +237,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 26
+           tanner 36
          </td>
          <td
            class=""
          >
            linsley
@@ -271,11 +271,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 27
+           tanner 37
          </td>
          <td
            class=""
          >
            linsley
@@ -305,11 +305,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 28
+           tanner 38
          </td>
          <td
            class=""
          >
            linsley
@@ -339,11 +339,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 29
+           tanner 39
          </td>
          <td
            class=""
          >
            linsley
@@ -373,11 +373,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 30
+           tanner 40
          </td>
          <td
            class=""
          >
            linsley
@@ -425,20 +425,20 @@
      </button>
       
      <span>
        Page 
        <strong>
-         3 of 10
+         4 of 10
        </strong>
         
      </span>
      <span>
        | Go to page: 
        <input
          style="width: 100px;"
          type="number"
-         value="3"
+         value="4"
        />
      </span>
       
      <select>
        <option
`;

exports[`renders a paginated table 2`] = `
Snapshot Diff:
- First value
+ Second value

@@ -67,11 +67,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 31
+           tanner 41
          </td>
          <td
            class=""
          >
            linsley
@@ -101,11 +101,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 32
+           tanner 42
          </td>
          <td
            class=""
          >
            linsley
@@ -135,11 +135,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 33
+           tanner 43
          </td>
          <td
            class=""
          >
            linsley
@@ -169,11 +169,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 34
+           tanner 44
          </td>
          <td
            class=""
          >
            linsley
@@ -203,11 +203,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 35
+           tanner 45
          </td>
          <td
            class=""
          >
            linsley
@@ -237,11 +237,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 36
+           tanner 46
          </td>
          <td
            class=""
          >
            linsley
@@ -271,11 +271,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 37
+           tanner 47
          </td>
          <td
            class=""
          >
            linsley
@@ -305,11 +305,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 38
+           tanner 48
          </td>
          <td
            class=""
          >
            linsley
@@ -339,11 +339,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 39
+           tanner 49
          </td>
          <td
            class=""
          >
            linsley
@@ -373,11 +373,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 40
+           tanner 50
          </td>
          <td
            class=""
          >
            linsley
@@ -425,20 +425,20 @@
      </button>
       
      <span>
        Page 
        <strong>
-         4 of 10
+         5 of 10
        </strong>
         
      </span>
      <span>
        | Go to page: 
        <input
          style="width: 100px;"
          type="number"
-         value="4"
+         value="5"
        />
      </span>
       
      <select>
        <option
`;

exports[`renders a paginated table 3`] = `
Snapshot Diff:
- First value
+ Second value

@@ -67,11 +67,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 41
+           tanner 91
          </td>
          <td
            class=""
          >
            linsley
@@ -101,11 +101,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 42
+           tanner 92
          </td>
          <td
            class=""
          >
            linsley
@@ -135,11 +135,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 43
+           tanner 93
          </td>
          <td
            class=""
          >
            linsley
@@ -169,11 +169,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 44
+           tanner 94
          </td>
          <td
            class=""
          >
            linsley
@@ -203,11 +203,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 45
+           tanner 95
          </td>
          <td
            class=""
          >
            linsley
@@ -237,11 +237,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 46
+           tanner 96
          </td>
          <td
            class=""
          >
            linsley
@@ -271,11 +271,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 47
+           tanner 97
          </td>
          <td
            class=""
          >
            linsley
@@ -305,11 +305,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 48
+           tanner 98
          </td>
          <td
            class=""
          >
            linsley
@@ -339,11 +339,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 49
+           tanner 99
          </td>
          <td
            class=""
          >
            linsley
@@ -373,11 +373,11 @@
          class=""
        >
          <td
            class=""
          >
-           tanner 50
+           tanner 100
          </td>
          <td
            class=""
          >
            linsley
@@ -414,31 +414,35 @@
       
      <button>
        &lt;
      </button>
       
-     <button>
+     <button
+       disabled=""
+     >
        &gt;
      </button>
       
-     <button>
+     <button
+       disabled=""
+     >
        &gt;&gt;
      </button>
       
      <span>
        Page 
        <strong>
-         5 of 10
+         10 of 10
        </strong>
         
      </span>
      <span>
        | Go to page: 
        <input
          style="width: 100px;"
          type="number"
-         value="5"
+         value="10"
        />
      </span>
       
      <select>
        <option
`;
